### PR TITLE
RequiredScripts Validation + additional input validation

### PR DESCRIPTION
**Pull request by yHSJ**
This PR introduces some new validations:

**NEW**
- **`InvalidScripts`**
   - **`InvalidScripts::MissingRequiredScriptWitnesses`**
       - Error if a script is required by an input but is not present in the reference inputs or witness set
   - **`InvalidScripts::ExtraneousScriptWitnesses`**
       - Error if there is a script witness present that is not required by an input

**MODIFIED**
- **`InvalidInputs`**
   - **`InvalidInputs::EmptyInputSet`**
      - Error if transaction has an empty input set (could this logic be moved to deserialization?)
   -  **`InvalidInputs::UnknownInput`**
       - Previously, this error was only thrown if an *input* was not found in the UTxO set. Now it will also be thrown if there is a *reference input* that is not found in the UTxO set.
       
       
**TODO**
- [x] Introduce new `scripts` logic
- [x] Introduce additional `inputs` logic
- [x] Tests for additional `inputs` logic
- [x] Tests for new `scripts` logic

## Summary by CodeRabbit

- **New Features**
  - Added validation for required script witnesses in transactions, including detection of missing or extraneous scripts.
  - Introduced support for script references in transaction outputs and improved script hash computation across multiple script types.
  - Transactions now explicitly track and expose required script hashes for validation.
  - Added minting policy witness requirements enforcement during transaction validation.
  - Extended governance proposal validation to require script witnesses for proposals containing script hashes.
- **Bug Fixes**
  - Transactions with empty input sets or unknown reference inputs are now correctly flagged as errors.
- **Tests**
  - Expanded test coverage with new and updated scenarios for script validation, empty inputs, reference input checks, minting policies, and valid/invalid witness cases.
  - Added comprehensive test data for complex transaction and script scenarios.
- **Documentation**
  - Enhanced inline documentation for new public interfaces and error variants related to script handling.
- **Chores**
  - Added error-level logging for invalid block validation results during ledger stage processing.

### CODE REVIEW

**Review by jeluard — APPROVED** [review]

**Comment by jeluard** [comment]
@yHSJ Is this ready?

**Comment by yHSJ** [comment]
> @yHSJ Is this ready?

Not just yet, no. I have to write tests for the new script logic, which will include changes to the JSON spec for `AssertValidationContext`

**Comment by yHSJ** [comment]
This is very much an incremental PR in that it introduces new functionality that "works"(TM), but is not ideal. We could spend a lot of time refactoring code to make this much nicer (i.e. we are looping over reference inputs and inputs several times, could move `provided_scripts` to the context), but I feel like that is part of the next iteration on this logic.

**Comment by yHSJ** [comment]
@jeluard Assuming the checks all pass, now this is ready to go :)

**Comment by jeluard** [comment]
@yHSJ `snapshots` tests fail with the following error:

```shell
Block invalid: Transaction { transaction_hash: Hash<32>("e974fecbf45ac386a76605e9e847a2e5d27c007fdd0be674cbad538e0c35fe01"), transaction_index: 0, violation: Scripts(ExtraneousScriptWitnesses([Hash<28>("fa24fb305126805cf2164c161d852a0e7330cf988f1fe558cf7d4a64")])) }
```

**Comment by yHSJ** [comment]
> This is very much an incremental PR in that it introduces new functionality that "works"(TM), but is not ideal. We could spend a lot of time refactoring code to make this much nicer (i.e. we are looping over reference inputs and inputs several times, could move `provided_scripts` to the context), but I feel like that is part of the next iteration on this logic.

Yes, forgot about proposal scripts. Fixing this AM

File: crates/amaru-kernel/src/serde_utils.rs [hunk 1]
```diff
@@ -65,7 +71,39 @@ impl From<TransactionOutputProxy> for TransactionOutput {
...
+    PlutusV3(Bytes),
+}
+
+impl HasProxy for ScriptRef {
+    type Proxy = ScriptRefProxy;
+}
+
+impl From<ScriptRefProxy> for ScriptRef {
+    // TODO: Is there a better way to do this? Just brute force to get it working for now...
+    #[allow(clippy::unwrap_used)]
```
Comment (jeluard) [inline crates/amaru-kernel/src/serde_utils.rs:97]: If it can fail, would be better to return a `Result` and handle it down the stream.

File: crates/amaru-ledger/src/rules/transaction/scripts.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                    $hash,
+                    "/",
+                    $variant,
+                    "/witness.cbor"
+                )),
+            )
+        };
+    }
+
+    #[test_case(fixture!("3b54f084af170b30565b1befe25860214a690a6c7a310e2902504dbc609c318e"); "happy path")]
```
Comment (jeluard) [inline crates/amaru-ledger/src/rules/transaction/scripts.rs:124]: How are all the json files generated? Would be nice to have some details/scripts shared.

File: crates/amaru/src/stages/ledger.rs [hunk 3]
```diff
@@ -115,6 +115,7 @@ impl<S: Store + Send, HS: HistoricalStores + Send> ValidateBlockStage<S, HS> {
         match rules::validate_block(&mut context, ProtocolParameters::default(), &block) {
             BlockValidation::Err(err) => return Err(err),
             BlockValidation::Invalid(err) => {
+                error!("Block invalid: {:?}", err);
```
Comment (jeluard) [inline crates/amaru/src/stages/ledger.rs:118]: Not sure it fully aligns with the current span driven approach. We might want to figure out a more global way to deal with error.

File: crates/amaru-kernel/src/lib.rs [hunk 4]
```diff
@@ -783,6 +788,63 @@ pub fn to_ex_units(witness_set: WitnessSet) -> ExUnits {
     }
 }
 
+/// Collect provided scripts and compute each ScriptHash in a witness set
+pub fn get_provided_scripts(witness_set: &MintedWitnessSet<'_>) -> BTreeSet<ScriptHash> {
+    let mut provided_scripts: BTreeSet<ScriptHash> = BTreeSet::new();
+    provided_scripts.extend(
```
Comment (jeluard) [inline crates/amaru-kernel/src/lib.rs:794]: Might be worth considering [chain](https://doc.rust-lang.org/std/iter/struct.Chain.html) as a way to reduce the overhead of multiple `extends` now probably not worth from a resource constraints here. More as a nice exercise towards a more functional approach :)

**Review by jeluard — COMMENTED** [review]
Nice! Would make sense to look at removing the todos(`network` access, refactoring?)  in a subsequent PR before we forget about them :)

File: crates/amaru-ledger/src/rules/transaction/scripts.rs (hunk 2 above)
Comment (yHSJ) [reply crates/amaru-ledger/src/rules/transaction/scripts.rs:124]: My fingers generate them 😛 

I find transactions on preprod that meet my basic needs, create the JSON files, then modify them to create different edge cases

File: crates/amaru/src/stages/ledger.rs (hunk 3 above)
Comment (yHSJ) [reply crates/amaru/src/stages/ledger.rs:118]: Just wanted to make sure something was getting printed for now.

File: crates/amaru-kernel/src/serde_utils.rs (hunk 1 above)
Comment (yHSJ) [reply crates/amaru-kernel/src/serde_utils.rs:97]: It's only something that ever runs in tests and it *should* as it's decoding test data which we've set up manually.

File: crates/amaru-ledger/src/rules/transaction/scripts.rs (hunk 2 above)
Comment (Quantumplation) [reply crates/amaru-ledger/src/rules/transaction/scripts.rs:124]: Also note that longer term, we'll flesh out the hand-rolled cases by [generating](https://github.com/SundaeSwap-finance/cardano-ledger-conformance-tests/tree/rrruko/dump-test-vectors) golden test vectors and saving them to [Cardano Blueprint](https://github.com/cardano-scaling/cardano-blueprint/pull/41)

The hand-rolled test cases here serve, effectively, as documentation that helps us think through the corner cases, and then then golden test vectors will help us ensure we have high coverage overall.
